File: CS/SchedulingDemo.Wpf/ViewModels/OutlookInspired/OutlookInspiredDemoViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Windows;
using Microsoft.Win32;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Xpf.Core;
using DevExpress.Xpf.Scheduling;
using DevExpress.Xpf.Scheduling.iCalendar;
using DevExpress.XtraScheduler;

namespace SchedulingDemo.ViewModels {
    public class OutlookInspiredDemoViewModel {
        static readonly IList<TimeSpan> TimeScales = new TimeSpan[] { TimeSpan.FromMinutes(60), TimeSpan.FromMinutes(30), TimeSpan.FromMinutes(15), TimeSpan.FromMinutes(10), TimeSpan.FromMinutes(6), TimeSpan.FromMinutes(5) };
        static readonly OutlookInspiredData data = new OutlookInspiredData();
        AppointmentItem selectedAppointment;

        protected OutlookInspiredDemoViewModel() {
            MinZoomFactor = 0;
            MaxZoomFactor = 100;
            ZoomStep = 20;
            ZoomFactor = 20;
            LinksViewModel = LinksViewModel.Create();
            ViewSettings = ViewSettingsViewModel.Create();
            PrintStyles = CalendarDetailPrintReportViewModel.Create();
            CalendarSettings = CalendarSettingsViewModel.Create();
        }

        public virtual IBackstageViewService BackstageViewService { get { return null; } }
        public virtual LinksViewModel LinksViewModel { get; set; }
        public virtual ViewSettingsViewModel ViewSettings { get; set; }
        public virtual CalendarDetailPrintReportViewModel PrintStyles { get; set; }
        public virtual CalendarSettingsViewModel CalendarSettings { get; set; }
        public virtual ObservableCollection<Calendar> Calendars { get { return data.Calendars; } }
        public virtual ObservableCollection<OutlookEvent> Events { get { return data.Events; } }
        public virtual ObservableCollection<Category> Categories { get { return data.Categories; } }

        [BindableProperty]
        public virtual AppointmentItem SelectedAppointment {
            get { return this.selectedAppointment; }
            set {
                if (this.selectedAppointment != null)
                    this.selectedAppointment.PropertyChanged -= PropertyChanged;
                AppointmentItem newSelectedAppointment = value;
                if (newSelectedAppointment != null)
                    newSelectedAppointment.PropertyChanged += PropertyChanged;
                this.selectedAppointment = newSelectedAppointment;
                UpdatePriority(this.selectedAppointment);
            }
        }

        public virtual TimeSpan TimeScale { get; set; }
        public virtual double ZoomFactor { get; set; }
        public virtual int MinZoomFactor { get; set; }
        public virtual int MaxZoomFactor { get; set; }
        public virtual bool HighImportanceChecked { get; set; }
        public virtual bool LowImportanceChecked { get; set; }

        int ZoomStep { get; set; }

        public void Print(SchedulerControl scheduler) {
            PrintingSettingsWindow printingSettingsWindow = new PrintingSettingsWindow();
            printingSettingsWindow.DataContext = PrintingSettingsWindowViewModel.Create(scheduler, Utils.GetReportTemplateCollection());
            printingSettingsWindow.Owner = Window.GetWindow(scheduler);
            printingSettingsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner;
            printingSettingsWindow.ShowDialog();
        }
        public void OutlookImport(SchedulerControl scheduler) {
            OutlookExchange(scheduler, OutlookExchangeType.Import);
        }
        public void OutlookExport(SchedulerControl scheduler) {
            OutlookExchange(scheduler, OutlookExchangeType.Export);
        }
        public void iCalendarImport(SchedulerControl scheduler) {
            ICalendarImporter importer = new ICalendarImporter(scheduler);
            importer.AppointmentItemImported += AppointmentItemImported;
            using(Stream stream = OpenRead("Calendar", "iCalendar files (*.ics)|*.ics")) {
                if (stream != null) 
                    importer.Import(stream);
            }
        }
        
        public void iCalendarExport(SchedulerControl scheduler) {
            ICalendarExporter exporter = new ICalendarExporter(scheduler);
            using(Stream stream = OpenWrite("Calendar", "iCalendar files (*.ics)|*.ics")) {
                if(stream != null) {
                    exporter.Export(stream);
                    stream.Flush();
                }
            }
        }
        public void Send(SchedulerControl scheduler) {
            Utils.SendMail(scheduler, scheduler.SelectedAppointments);
        }
        public void SetHighPriority() {
            if (SelectedAppointment == null)
                return;
            EventPriority priority = (EventPriority)SelectedAppointment.CustomFields["Priority"];
            SelectedAppointment.CustomFields["Priority"] = priority == EventPriority.HighImportance ? EventPriority.None : EventPriority.HighImportance;
            if(LowImportanceChecked)
                LowImportanceChecked = !HighImportanceChecked;
        }
        public void SetLowPriority() {
            if (SelectedAppointment == null)
                return;
            EventPriority priority = (EventPriority)SelectedAppointment.CustomFields["Priority"];
            SelectedAppointment.CustomFields["Priority"] = priority == EventPriority.LowImportance ? EventPriority.None : EventPriority.LowImportance;
            if (HighImportanceChecked)
                HighImportanceChecked = !LowImportanceChecked;
        }
        public void OpenViewSettingsWindow(SchedulerControl scheduler) {
            CalendarSettingsWindow settingsWindow = new CalendarSettingsWindow();
            CalendarSettingsViewModel newCalendarSettings = CalendarSettingsViewModel.Create();
            newCalendarSettings.ApplySettings(CalendarSettings);
            settingsWindow.DataContext = newCalendarSettings;
            settingsWindow.Owner = Window.GetWindow(scheduler);
            settingsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner;
            if (settingsWindow.ShowDialog() == true)
                CalendarSettings.ApplySettings(newCalendarSettings);

        }
        public void ResetViewSettings() {
            CalendarSettings = CalendarSettingsViewModel.Create();
        }
        public void CloseBackstageView() {
            BackstageViewService.Close();
        }

        protected void OnZoomFactorChanged() {
            int value = (int)(ZoomFactor / ZoomStep);
            TimeScale = TimeScales[value];
        }
        protected void OnTimeScaleChanged() {
            int value = TimeScales.IndexOf(TimeScale);
            ZoomFactor = ZoomStep * value;
        }

        void UpdatePriority(AppointmentItem appointment) {
            if (appointment == null)
                return;
            EventPriority priority = (EventPriority)appointment.CustomFields["Priority"];
            HighImportanceChecked = priority == EventPriority.HighImportance;
            LowImportanceChecked = priority == EventPriority.LowImportance;
        }
        void PropertyChanged(object sender, PropertyChangedEventArgs e) {
            if (e.PropertyName == "Priority")
                UpdatePriority(sender as AppointmentItem);
        }

        Stream OpenRead(string fileName, string filter) {
            OpenFileDialog dialog = new OpenFileDialog() { FileName = fileName, Filter = filter, FilterIndex = 1 };
            if(dialog.ShowDialog() != true)
                return null;
            return dialog.OpenFile();
        }
        Stream OpenWrite(string fileName, string filter) {
            SaveFileDialog dialog = new SaveFileDialog() { FileName = fileName, Filter = filter, FilterIndex = 1 };
            if(dialog.ShowDialog() != true)
                return null;
            return dialog.OpenFile();
        }
        void OutlookExchange(SchedulerControl scheduler, OutlookExchangeType exchangeType) {
            try {
                string[] outlookCalendarPaths = DevExpress.XtraScheduler.Outlook.OutlookExchangeHelper.GetOutlookCalendarPaths();
                if (outlookCalendarPaths == null || outlookCalendarPaths.Length == 0)
                    return;

                OutlookExchangeOptionsWindow optionsWindow = new OutlookExchangeOptionsWindow();
                optionsWindow.DataContext = OutlookExchangeOptionsWindowViewModel.Create(scheduler, exchangeType, outlookCalendarPaths);
                optionsWindow.Owner = Window.GetWindow(scheduler);
                optionsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner;
                optionsWindow.ShowDialog();
            }
            catch {
                DXMessageBox.Show(String.Format("Unable to {0}.\nCheck whether MS Outlook is installed.", "get the list of available calendars from Microsoft Outlook"),
                    "Import from MS Outlook", MessageBoxButton.OK, MessageBoxImage.Warning);
            }           
        }

        void AppointmentItemImported(object sender, AppointmentItemImportedEventArgs e) {
            CustomFieldCollection customFields = e.Appointment.CustomFields;
            customFields["Priority"] = ConvertToEventPriority(customFields["Priority"]);
            customFields["IsPrivate"] = ConvertToBoolean(customFields["IsPrivate"]);
        }

        bool ConvertToBoolean(object customField) {
            return customField == null ? false : AreStringsEqual(customField.ToString(), "true");
        }

        EventPriority ConvertToEventPriority(object customField) {
            if (customField == null)
                return EventPriority.None;
            EventPriority priority = EventPriority.HighImportance;
            string value = customField.ToString();
            if (AreStringsEqual(value, priority.ToString()))
                return priority;
            priority = EventPriority.LowImportance;
            if (AreStringsEqual(value, priority.ToString()))
                return priority;
            return EventPriority.None;
        }

        bool AreStringsEqual(string left, string right) {
            return String.Compare(left, right, StringComparison.InvariantCultureIgnoreCase) == 0;
        }
    }
}

File: VB/SchedulingDemo.Wpf/ViewModels/OutlookInspired/OutlookInspiredDemoViewModel.vb
Imports System
Imports System.Collections.ObjectModel
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Windows
Imports Microsoft.Win32
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Xpf.Core
Imports DevExpress.Xpf.Scheduling
Imports DevExpress.Xpf.Scheduling.iCalendar
Imports DevExpress.XtraScheduler

Namespace SchedulingDemo.ViewModels
    Public Class OutlookInspiredDemoViewModel
        Private Shared ReadOnly TimeScales As IList(Of TimeSpan) = New TimeSpan() { TimeSpan.FromMinutes(60), TimeSpan.FromMinutes(30), TimeSpan.FromMinutes(15), TimeSpan.FromMinutes(10), TimeSpan.FromMinutes(6), TimeSpan.FromMinutes(5) }
        Private Shared ReadOnly data As New OutlookInspiredData()

        Private selectedAppointment_Renamed As AppointmentItem

        Protected Sub New()
            MinZoomFactor = 0
            MaxZoomFactor = 100
            ZoomStep = 20
            ZoomFactor = 20
            LinksViewModel = LinksViewModel.Create()
            ViewSettings = ViewSettingsViewModel.Create()
            PrintStyles = CalendarDetailPrintReportViewModel.Create()
            CalendarSettings = CalendarSettingsViewModel.Create()
        End Sub

        Public Overridable ReadOnly Property BackstageViewService() As IBackstageViewService
            Get
                Return Nothing
            End Get
        End Property
        Public Overridable Property LinksViewModel() As LinksViewModel
        Public Overridable Property ViewSettings() As ViewSettingsViewModel
        Public Overridable Property PrintStyles() As CalendarDetailPrintReportViewModel
        Public Overridable Property CalendarSettings() As CalendarSettingsViewModel
        Public Overridable ReadOnly Property Calendars() As ObservableCollection(Of Calendar)
            Get
                Return data.Calendars
            End Get
        End Property
        Public Overridable ReadOnly Property Events() As ObservableCollection(Of OutlookEvent)
            Get
                Return data.Events
            End Get
        End Property
        Public Overridable ReadOnly Property Categories() As ObservableCollection(Of Category)
            Get
                Return data.Categories
            End Get
        End Property

        <BindableProperty>
        Public Overridable Property SelectedAppointment() As AppointmentItem
            Get
                Return Me.selectedAppointment_Renamed
            End Get
            Set(ByVal value As AppointmentItem)
                If Me.selectedAppointment_Renamed IsNot Nothing Then
                    RemoveHandler Me.selectedAppointment_Renamed.PropertyChanged, AddressOf PropertyChanged
                End If
                Dim newSelectedAppointment As AppointmentItem = value
                If newSelectedAppointment IsNot Nothing Then
                    AddHandler newSelectedAppointment.PropertyChanged, AddressOf PropertyChanged
                End If
                Me.selectedAppointment_Renamed = newSelectedAppointment
                UpdatePriority(Me.selectedAppointment_Renamed)
            End Set
        End Property

        Public Overridable Property TimeScale() As TimeSpan
        Public Overridable Property ZoomFactor() As Double
        Public Overridable Property MinZoomFactor() As Integer
        Public Overridable Property MaxZoomFactor() As Integer
        Public Overridable Property HighImportanceChecked() As Boolean
        Public Overridable Property LowImportanceChecked() As Boolean

        Private Property ZoomStep() As Integer

        Public Sub Print(ByVal scheduler As SchedulerControl)
            Dim printingSettingsWindow As New PrintingSettingsWindow()
            printingSettingsWindow.DataContext = PrintingSettingsWindowViewModel.Create(scheduler, Utils.GetReportTemplateCollection())
            printingSettingsWindow.Owner = Window.GetWindow(scheduler)
            printingSettingsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner
            printingSettingsWindow.ShowDialog()
        End Sub
        Public Sub OutlookImport(ByVal scheduler As SchedulerControl)
            OutlookExchange(scheduler, OutlookExchangeType.Import)
        End Sub
        Public Sub OutlookExport(ByVal scheduler As SchedulerControl)
            OutlookExchange(scheduler, OutlookExchangeType.Export)
        End Sub
        Public Sub iCalendarImport(ByVal scheduler As SchedulerControl)
            Dim importer As New ICalendarImporter(scheduler)
            AddHandler importer.AppointmentItemImported, AddressOf AppointmentItemImported
            Using stream As Stream = OpenRead("Calendar", "iCalendar files (*.ics)|*.ics")
                If stream IsNot Nothing Then
                    importer.Import(stream)
                End If
            End Using
        End Sub

        Public Sub iCalendarExport(ByVal scheduler As SchedulerControl)
            Dim exporter As New ICalendarExporter(scheduler)
            Using stream As Stream = OpenWrite("Calendar", "iCalendar files (*.ics)|*.ics")
                If stream IsNot Nothing Then
                    exporter.Export(stream)
                    stream.Flush()
                End If
            End Using
        End Sub
        Public Sub Send(ByVal scheduler As SchedulerControl)
            Utils.SendMail(scheduler, scheduler.SelectedAppointments)
        End Sub
        Public Sub SetHighPriority()
            If SelectedAppointment Is Nothing Then
                Return
            End If
            Dim priority As EventPriority = CType(SelectedAppointment.CustomFields("Priority"), EventPriority)
            SelectedAppointment.CustomFields("Priority") = If(priority = EventPriority.HighImportance, EventPriority.None, EventPriority.HighImportance)
            If LowImportanceChecked Then
                LowImportanceChecked = Not HighImportanceChecked
            End If
        End Sub
        Public Sub SetLowPriority()
            If SelectedAppointment Is Nothing Then
                Return
            End If
            Dim priority As EventPriority = CType(SelectedAppointment.CustomFields("Priority"), EventPriority)
            SelectedAppointment.CustomFields("Priority") = If(priority = EventPriority.LowImportance, EventPriority.None, EventPriority.LowImportance)
            If HighImportanceChecked Then
                HighImportanceChecked = Not LowImportanceChecked
            End If
        End Sub
        Public Sub OpenViewSettingsWindow(ByVal scheduler As SchedulerControl)
            Dim settingsWindow As New CalendarSettingsWindow()
            Dim newCalendarSettings As CalendarSettingsViewModel = CalendarSettingsViewModel.Create()
            newCalendarSettings.ApplySettings(CalendarSettings)
            settingsWindow.DataContext = newCalendarSettings
            settingsWindow.Owner = Window.GetWindow(scheduler)
            settingsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner
            If settingsWindow.ShowDialog() = True Then
                CalendarSettings.ApplySettings(newCalendarSettings)
            End If

        End Sub
        Public Sub ResetViewSettings()
            CalendarSettings = CalendarSettingsViewModel.Create()
        End Sub
        Public Sub CloseBackstageView()
            BackstageViewService.Close()
        End Sub

        Protected Sub OnZoomFactorChanged()
            Dim value As Integer = CInt((ZoomFactor / ZoomStep))
            TimeScale = TimeScales(value)
        End Sub
        Protected Sub OnTimeScaleChanged()
            Dim value As Integer = TimeScales.IndexOf(TimeScale)
            ZoomFactor = ZoomStep * value
        End Sub

        Private Sub UpdatePriority(ByVal appointment As AppointmentItem)
            If appointment Is Nothing Then
                Return
            End If
            Dim priority As EventPriority = CType(appointment.CustomFields("Priority"), EventPriority)
            HighImportanceChecked = priority = EventPriority.HighImportance
            LowImportanceChecked = priority = EventPriority.LowImportance
        End Sub
        Private Sub PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
            If e.PropertyName = "Priority" Then
                UpdatePriority(TryCast(sender, AppointmentItem))
            End If
        End Sub

        Private Function OpenRead(ByVal fileName As String, ByVal filter As String) As Stream
            Dim dialog As New OpenFileDialog() With {.FileName = fileName, .Filter = filter, .FilterIndex = 1}
            If Not dialog.ShowDialog().Equals(True) Then
                Return Nothing
            End If
            Return dialog.OpenFile()
        End Function
        Private Function OpenWrite(ByVal fileName As String, ByVal filter As String) As Stream
            Dim dialog As New SaveFileDialog() With {.FileName = fileName, .Filter = filter, .FilterIndex = 1}
            If Not dialog.ShowDialog().Equals(True) Then
                Return Nothing
            End If
            Return dialog.OpenFile()
        End Function
        Private Sub OutlookExchange(ByVal scheduler As SchedulerControl, ByVal exchangeType As OutlookExchangeType)
            Try
                Dim outlookCalendarPaths() As String = DevExpress.XtraScheduler.Outlook.OutlookExchangeHelper.GetOutlookCalendarPaths()
                If outlookCalendarPaths Is Nothing OrElse outlookCalendarPaths.Length = 0 Then
                    Return
                End If

                Dim optionsWindow As New OutlookExchangeOptionsWindow()
                optionsWindow.DataContext = OutlookExchangeOptionsWindowViewModel.Create(scheduler, exchangeType, outlookCalendarPaths)
                optionsWindow.Owner = Window.GetWindow(scheduler)
                optionsWindow.WindowStartupLocation = WindowStartupLocation.CenterOwner
                optionsWindow.ShowDialog()
            Catch
                DXMessageBox.Show(String.Format("Unable to {0}." & ControlChars.Lf & "Check whether MS Outlook is installed.", "get the list of available calendars from Microsoft Outlook"), "Import from MS Outlook", MessageBoxButton.OK, MessageBoxImage.Warning)
            End Try
        End Sub

        Private Sub AppointmentItemImported(ByVal sender As Object, ByVal e As AppointmentItemImportedEventArgs)
            Dim customFields As CustomFieldCollection = e.Appointment.CustomFields
            customFields("Priority") = ConvertToEventPriority(customFields("Priority"))
            customFields("IsPrivate") = ConvertToBoolean(customFields("IsPrivate"))
        End Sub

        Private Function ConvertToBoolean(ByVal customField As Object) As Boolean
            Return If(customField Is Nothing, False, AreStringsEqual(customField.ToString(), "true"))
        End Function

        Private Function ConvertToEventPriority(ByVal customField As Object) As EventPriority
            If customField Is Nothing Then
                Return EventPriority.None
            End If
            Dim priority As EventPriority = EventPriority.HighImportance
            Dim value As String = customField.ToString()
            If AreStringsEqual(value, priority.ToString()) Then
                Return priority
            End If
            priority = EventPriority.LowImportance
            If AreStringsEqual(value, priority.ToString()) Then
                Return priority
            End If
            Return EventPriority.None
        End Function

        Private Function AreStringsEqual(ByVal left As String, ByVal right As String) As Boolean
            Return String.Compare(left, right, StringComparison.InvariantCultureIgnoreCase) = 0
        End Function
    End Class
End Namespace
